routing/routing_session.cpp

```diff
@@ -786,6 +724,101 @@ void RoutingSession::CopyTraffic(traffic::AllMwmTrafficInfo & trafficColoring) c
   TrafficCache::CopyTraffic(trafficColoring);
 }
 
+void RoutingSession::ProcessSpeedCameras(GpsInfo const & info)
+{
+  CHECK_THREAD_CHECKER(m_threadChecker, ());
+
+  auto const passedDistanceMeters = m_route->GetCurrentDistanceFromBeginMeters();
+  auto const & segments = m_route->GetRouteSegments();
+
+  // Step 1. Find new cameras and cache them.
+  size_t firstNotChecked = m_firstNotCheckedSpeedCameraIndex;
+  if (firstNotChecked == segments.size())
+    return;
+
+  CHECK_LESS(firstNotChecked, segments.size(), ());
+
+  double distToPrevSegment = segments[firstNotChecked].GetDistFromBeginningMeters();
+  double distFromCurPosToLatestCheckedSegmentM = distToPrevSegment - passedDistanceMeters;
+
+  while (firstNotChecked < segments.size() &&
+         distFromCurPosToLatestCheckedSegmentM < SpeedCameraOnRoute::kLookAheadDistanceMeters)
+  {
+    auto const & lastSegment = segments[firstNotChecked];
+    auto const & speedCamsVector = lastSegment.GetSpeedCams();
+    double segmentLength = m_route->GetSegLenMeters(firstNotChecked);
+
+    for (auto const & speedCam : speedCamsVector)
+    {
+      segmentLength *= speedCam.m_coef;
+      m_cachedSpeedCameras.emplace(distToPrevSegment + segmentLength, speedCam.m_maxSpeedKmPH);
+    }
+
+    distToPrevSegment = lastSegment.GetDistFromBeginningMeters();
+    distFromCurPosToLatestCheckedSegmentM = distToPrevSegment - passedDistanceMeters;
+    ++firstNotChecked;
+  }
+
+  m_firstNotCheckedSpeedCameraIndex = firstNotChecked;
+
+  // Step 2. Process warned cameras.
+  while (!m_warnedSpeedCameras.empty())
+  {
+    auto const & oldestCamera = m_warnedSpeedCameras.front();
+    double const distBetweenCameraAndCurrentPos = passedDistanceMeters - oldestCamera.m_distFromBeginMeters;
+    if (distBetweenCameraAndCurrentPos < SpeedCameraOnRoute::kInfluenceZoneMeters)
+      break;
+
+    m_warnedSpeedCameras.pop();
+  }
+
+  // We will turn off warning in UI only after all cameras, about which
+  // we were be notified, exhausted.
```

Comment:
s/were be/were/ or s/were be/have been/
s/exhausted/are exhausted/

commas are superfluous

Reply:
done